### ISSUE TITLE
A few questions about the code

### ISSUE THREAD

**Issue by Minorant** [1zhou-Wang/MemVR#4]
Thank you for your wonderful work! I have a few questions about this code.
  1.  Why is the ratio of the mean of the weights in the linear layer of the FFN to the mean of the visual tokens used as a scaling factor for the visual tokens? I’m curious about the reasoning for this approach.
https://github.com/1zhou-Wang/MemVR/blob/737d05045e65d77465cb425df1112c17d6797b1f/memvr.py#L244-L245
  2.  Why is the entropy calculated by directly considering the top-k largest logits? Doesn't this mean that the words selected for the entropy at each layer are not corresponding to the same vocabulary?
https://github.com/1zhou-Wang/MemVR/blob/737d05045e65d77465cb425df1112c17d6797b1f/memvr.py#L212-L215
Thanks in advance for your time and help!

**Comment by 1zhou-Wang** [comment]
Hi,
Glad to share my view with you!

_1.  Why is the ratio of the mean of the weights in the linear layer of the FFN to the mean of the visual tokens used as a scaling factor for the visual tokens? I’m curious about the reasoning for this approach._
This part of calculation serves for the MemVR Adaptor, where the scaled visual feature has been used. Since the Adaptor's output will be added to the regular FFN output with specific ratio that is called 'Retrace Ratio', we would like to keep the final result's range as close to the original one as it can, as either higher or lower value would disturb the reasoning of LLM. And that's why the visual feature has been scaled according to the value of certain FFN weights --- we want to keep the calculation results from the adaptor as close to the value of FFN as it can! 

_2. Why is the entropy calculated by directly considering the top-k largest logits? Doesn't this mean that the words selected for the entropy at each layer are not corresponding to the same vocabulary?_
Sure that the entropy at each hidden-layer is not, indeed, calculated by collecting probabilities of the same group of predictions.  This is because the value of entropy represents for the 'uncertainty' of the hidden-layers, but not how the logits change from pre-mature-layers (which are hidden-layers) to mature-layer (which is often the last layer). In other words, we use this value to describe if the model has a preference among all logits, where lower value indicates more confident reasoning process even the early-state preference would not be the one to be predicted at the last layer. 

I hope that solves your puzzle. Do let me know if you have further concerns or questions!

**Comment by 1zhou-Wang** [comment]
Hi Young,

Top 10 logits selection is sufficient for layer entropy calculation, as keep increasing k will not make significant difference in the result. And of course, you're welcome to conduct experiments on the effect of increasing the number of k on MemVR's evaluation results.

Yizhou
